webapp/handlers.py

```diff
@@ -0,0 +1,94 @@
+import socket
+from urllib.parse import unquote, urlparse, urlunparse
+
+import flask
+import webapp.template_utils as template_utils
+
+
+def set_handlers(app):
+    @app.context_processor
+    def utility_processor():
+        """
+        This defines the set of properties and functions that will be added
+        to the default context for processing templates. All these items
+        can be used in all templates
+        """
+
+        return {
+            # Functions
+            "format_date": template_utils.format_date,
+            "get_json_feed": template_utils.get_json_feed_content,
+            "get_year": template_utils.get_year,
+            "replace_admin": template_utils.replace_admin,
+            "versioned_static": template_utils.versioned_static,
+            "truncate_chars": template_utils.truncate_chars,
+        }
+
+    # Error handlers
+    # ===
+    @app.errorhandler(404)
+    def page_not_found(error):
+        """
+        For 404 pages, display the 404.html template,
+        passing through the error description.
+        """
+
+        return flask.render_template("404.html", error=error.description), 404
+
+    @app.errorhandler(500)
+    @app.errorhandler(501)
+    @app.errorhandler(502)
+    @app.errorhandler(503)
+    @app.errorhandler(504)
+    @app.errorhandler(505)
+    def internal_error(error):
+        error_name = getattr(error, "name", type(error).__name__)
+        return_code = getattr(error, "code", 500)
+
+        return (
+            flask.render_template("500.html", error_name=error_name),
+            return_code,
+        )
+
+    # Global tasks for all requests
+    # ===
+    @app.before_request
+    def clear_trailing():
+        """
+        Remove trailing slashes from all routes
+        We like our URLs without slashes
+        """
+
+        parsed_url = urlparse(unquote(flask.request.url))
+        path = parsed_url.path
+
+        if path != "/" and path.endswith("/"):
+            new_uri = urlunparse(parsed_url._replace(path=path[:-1]))
+
+            return flask.redirect(new_uri)
+
+    @app.after_request
+    def add_headers(response):
```

Comment:
I don't think this is needed here, this is more specific for apps that have sessions

Reply:
hm, what about the cache-control though?